ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,668 @@
+Finally, processing starts again at `L12`, during which the Process BPMN element
+of workflow `1` in workflow instance `2` transitions to `Element_Activated`.
+
+Now, let's make an imaginary bug fix in the key generation. From now, only odd
+numbers are generated.
+
+Updating to this new version from `S1` with reprocessing starting at `L10`,
+results in a workflow instance with key `3` in the state. Then, resprocessing
+`L11` is skipped again (there is still no procesor for this record). And lastly,
+when processing `L12`, there is a failure, because the workflow instance with
+key `2` cannot be found.
+
+**In the engine proposed by this document**, the state of `S1` is copied and
+replay starts at `L10`. We need to note additionally, that `L10` and `L12` are
+commands, while `L11` is an event.
+
+First, `L10` is skipped, because commands are skipped during replay. Then,
+during the replay of `L11`, a workflow instance is stored in the state with key
+`2` as described by the data inside `L11`. Finally, processing starts again at
+`L12`, during which the Process BPMN element of workflow `1` in workflow
+instance `2` transitions to `Element_Activated`.
+
+Now, let's again make the same imaginary bug fix in the key generation. From
+now, only odd numbers are generated.
+
+Updating to this new version from `S1` with replay starting at `L10` results is
+the same state as before the update. First `L10` is skipped, because commands
+are skipped during replay. Then, during the replay of `L11`, a workflow instance
+is stored in the state with key `2` as described by the data inside `L11`. Note,
+that this key is not newly generated, in which case it would've had the value
+`3`. Finally, processing starts again at `L12`, during which the Process BPMN
+element of workflow `1` in workflow instance `2` transitions to
+`Element_Activated`.
+
+Note, that in this case the workflow instance was correctly referred to from
+`L12`, because the events were directly applied as changes to the state instead of reapplying the business logic (which had changed).
+
+This example also shows that the effect of this proposal is that bug fixes in
+the form of business logic changes can no longer be applied to the processing of
+records that have already been processed previously. Their situation will not be
+altered since their records have already been written to the log stream in the
+form of events.
+ 
+## Key Generation
+
+The failure in the previous example occurred in key generation. Even though this
+proposal improves dealing with changes to keys, we should consider keys in more
+detail.
+
+A key (i.e. a long value) identifies an entity (e.g. a workflow instance) uniquely across all partitions. Most records have a key, except for user commands that don't refer to an existing resource (e.g. create workflow instance command). When a record is written for a new entity when processing a command, then a new key is generated.
```

Comment:
Keys really are unique over all partitions, because they are prefixed by the partition id. Have a look at `Protocol.encodePartitionId(int, long)` which is used by KeyState. 

Records always discuss log stream records. I don't think we discuss database entries as records, but I could be mistaken. At least in this document we always mean log stream records.

I've also tried to explain the key generation a little more detailed. Please have a look.

Reply:
I find it still confusing. For the discussion in this document it might suffice to talk about unique-per partition keys. The thing that threw me off was this sentence "uniquely across all partitions". 

Not knowing how the cross-parition keys are generated I was thinking that there is one key generator for all partitions. And then I was wondering: Where do we store the latest key for a partition independent key generator?

Knowing now that the key gets a partition prefix and we have one key generator per partition, I think I understand it now.

I am torn between explaining key syntax better or just dropping the cross partition uniqueness altogether. Most of the document talks about stuff that happens within a partition. 

Feel free to leave it as is, or make changes as you see fit.

Reply:
I don't think unique-per partition is fitting, since its even more unique than that (it really is unique across all partitions). But I agree that we might be able to drop it all together, since I think it doesn't add anything for this proposal.

@saig0 I kinda agree that it doesn't really matter for this proposal whether or not keys identify entities across partitions or not. I think what matters is providing a short description of what keys are. For this proposal it might be enough to define keys as: uniquely identifies entities (e.g. a workflow instance). WDYT?

Reply:
We've decided to remove the mention of partitions in key generation paragraph.